dali/pipeline/operators/sequence/sequence_rearrange.h

```diff
@@ -0,0 +1,65 @@
+// Copyright (c) 2019, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_PIPELINE_OPERATORS_SEQUENCE_SEQUENCE_REARRANGE_H_
+#define DALI_PIPELINE_OPERATORS_SEQUENCE_SEQUENCE_REARRANGE_H_
+
+#include <tuple>
+#include <vector>
+
+#include "dali/pipeline/operators/operator.h"
+
+namespace dali {
+
+static Index GetSeqLength(const std::vector<Index>& seq_shape) {
+  return seq_shape[0];
+}
+
+static std::tuple<std::vector<Index>, Index> GetNewShapeAndElementSize(
+    const std::vector<Index>& in_sample_shape, const std::vector<int>& new_order) {
+  const int in_seq_length = GetSeqLength(in_sample_shape);
+  for (const auto& src_idx : new_order) {
+    DALI_ENFORCE(0 <= src_idx && src_idx < in_seq_length,
+                 "Source element src_idx must be in [0, input_sequence_length = " +
+                     std::to_string(in_seq_length) +
+                     ") for new_order argument, but it is: " + std::to_string(src_idx));
+  }
+  Index element_size = Product(in_sample_shape) / GetSeqLength(in_sample_shape);
+
+  const int out_seq_length = new_order.size();
+  auto new_sample_shape = in_sample_shape;
+  new_sample_shape[0] = out_seq_length;
+  return std::tuple<std::vector<Index>, Index>{new_sample_shape, element_size};
+}
+
+template <typename Backend>
+class SequenceRearrange : public Operator<Backend> {
+ public:
+  inline explicit SequenceRearrange(const OpSpec& spec)
+      : Operator<Backend>(spec), new_order_(spec.GetRepeatedArgument<int>("new_order")) {
```

Comment:
`inline` is redundant for functions defined inside class declaration:

> A function defined entirely inside a class/struct/union definition, whether it's a member function or a non-member friend function, is implicitly an inline function. 

https://en.cppreference.com/w/cpp/language/inline

Reply:
It is, but it is also used to point out the intention. If someone decide to move it to cc file then it would be clear if that function should stay inlined or not.

Reply:
It must be defined in a header because it's a template.
Inline here is of limited utility, since we're calling expensive functions (GetRepeatedArgument).